File: apps/api/internal/domain/port/repository.go
package port

import (
	"context"

	"github.com/neurodynecorp/api/internal/domain/entity"
)

// UserRepository defines the driven port for user persistence.
type UserRepository interface {
	Create(ctx context.Context, user *entity.User) error
	GetByID(ctx context.Context, id string) (*entity.User, error)
	GetByEmail(ctx context.Context, email string) (*entity.User, error)
	Update(ctx context.Context, user *entity.User) error
	Delete(ctx context.Context, id string) error
	List(ctx context.Context, filter UserFilter) ([]*entity.User, int64, error)
}

type UserFilter struct {
	Role     entity.Role
	IsActive *bool
	Search   string
	Page     int
	PageSize int
}

// ProjectRepository defines the driven port for project persistence.
type ProjectRepository interface {
	Create(ctx context.Context, project *entity.Project) error
	GetByID(ctx context.Context, id string) (*entity.Project, error)
	Update(ctx context.Context, project *entity.Project) error
	Delete(ctx context.Context, id string) error
	ListByClient(ctx context.Context, clientID string, page, pageSize int) ([]*entity.Project, int64, error)
	ListByStatus(ctx context.Context, status entity.ProjectStatus, page, pageSize int) ([]*entity.Project, int64, error)
	List(ctx context.Context, filter ProjectFilter) ([]*entity.Project, int64, error)
}

type ProjectFilter struct {
	ClientID string
	Status   entity.ProjectStatus
	Type     entity.ProjectType
	Search   string
	Page     int
	PageSize int
}

// SpecificationRepository defines the driven port for specification persistence.
type SpecificationRepository interface {
	Create(ctx context.Context, spec *entity.Specification) error
	GetByID(ctx context.Context, id string) (*entity.Specification, error)
	GetByProjectID(ctx context.Context, projectID string) (*entity.Specification, error)
	Update(ctx context.Context, spec *entity.Specification) error
	ListVersions(ctx context.Context, projectID string) ([]*entity.Specification, error)
}

// QuestionRepository defines the driven port for questionnaire persistence.
type QuestionRepository interface {
	GetAll(ctx context.Context) ([]*entity.Question, error)
	GetByCategory(ctx context.Context, category string) ([]*entity.Question, error)
	GetByID(ctx context.Context, id string) (*entity.Question, error)
	Upsert(ctx context.Context, question *entity.Question) error
}

// QuestionnaireResponseRepository defines the driven port for questionnaire responses.
type QuestionnaireResponseRepository interface {
	Create(ctx context.Context, response *entity.QuestionnaireResponse) error
	GetByID(ctx context.Context, id string) (*entity.QuestionnaireResponse, error)
	GetByProjectID(ctx context.Context, projectID string) (*entity.QuestionnaireResponse, error)
	Update(ctx context.Context, response *entity.QuestionnaireResponse) error
}

// InvoiceRepository defines the driven port for invoice persistence.
type InvoiceRepository interface {
	Create(ctx context.Context, invoice *entity.Invoice) error
	GetByID(ctx context.Context, id string) (*entity.Invoice, error)
	ListByClient(ctx context.Context, clientID string, page, pageSize int) ([]*entity.Invoice, int64, error)
	ListByProject(ctx context.Context, projectID string) ([]*entity.Invoice, error)
	Update(ctx context.Context, invoice *entity.Invoice) error
}

// TaskRepository defines the driven port for task persistence.
type TaskRepository interface {
	Create(ctx context.Context, task *entity.Task) error
	GetByID(ctx context.Context, id string) (*entity.Task, error)
	Update(ctx context.Context, task *entity.Task) error
	Delete(ctx context.Context, id string) error
	ListByProject(ctx context.Context, projectID string, filter TaskFilter) ([]*entity.Task, int64, error)
	ListBySprint(ctx context.Context, sprintID string) ([]*entity.Task, error)
}

type TaskFilter struct {
	Status     entity.TaskStatus
	AssigneeID string
	Priority   entity.TaskPriority
	Page       int
	PageSize   int
}

// SprintRepository defines the driven port for sprint persistence.
type SprintRepository interface {
	Create(ctx context.Context, sprint *entity.Sprint) error
	GetByID(ctx context.Context, id string) (*entity.Sprint, error)
	Update(ctx context.Context, sprint *entity.Sprint) error
	ListByProject(ctx context.Context, projectID string) ([]*entity.Sprint, error)
	GetActive(ctx context.Context, projectID string) (*entity.Sprint, error)
}

// NotificationRepository defines the driven port for notification persistence.
type NotificationRepository interface {
	Create(ctx context.Context, notification *entity.Notification) error
	GetByUser(ctx context.Context, userID string, page, pageSize int) ([]*entity.Notification, int64, error)
	MarkAsRead(ctx context.Context, id string) error
	MarkAllAsRead(ctx context.Context, userID string) error
	CountUnread(ctx context.Context, userID string) (int64, error)
}

// MessageRepository defines the driven port for message persistence.
type MessageRepository interface {
	Create(ctx context.Context, message *entity.Message) error
	GetByThread(ctx context.Context, threadID string, page, pageSize int) ([]*entity.Message, int64, error)
	ListByProject(ctx context.Context, projectID string, page, pageSize int) ([]*entity.Message, int64, error)
}

// ThreadRepository defines the driven port for thread persistence.
type ThreadRepository interface {
	Create(ctx context.Context, thread *entity.Thread) error
	GetByID(ctx context.Context, id string) (*entity.Thread, error)
	ListByProject(ctx context.Context, projectID string) ([]*entity.Thread, error)
}

File: apps/server/src/domain/port/repository.ts
import type {
  User,
  Project,
  Specification,
  Question,
  QuestionnaireResponse,
  Invoice,
  Task,
  Sprint,
  Notification,
  Message,
  Thread,
  ProjectStatus,
  ProjectType,
  TaskStatus,
  TaskPriority,
} from "../entity/index.js";
import type { Role as UserRole } from "../entity/user.js";
import type { RBACRole } from "../entity/role.js";

// ── Filters ──────────────────────────────────────────────────────────────────

export interface UserFilter {
  role?: UserRole;
  isActive?: boolean;
  company?: string;
  search?: string;
}

export interface ProjectFilter {
  clientId?: string;
  status?: ProjectStatus;
  type?: ProjectType;
  assignedTeamMember?: string;
  search?: string;
}

export interface TaskFilter {
  projectId?: string;
  sprintId?: string;
  status?: TaskStatus;
  priority?: TaskPriority;
  assigneeId?: string;
  reporterId?: string;
  search?: string;
}

// ── Repository Interfaces ────────────────────────────────────────────────────

export interface UserRepository {
  findById(id: string): Promise<User | null>;
  findByEmail(email: string): Promise<User | null>;
  findAll(filter?: UserFilter): Promise<User[]>;
  create(user: User): Promise<User>;
  update(user: User): Promise<User>;
  delete(id: string): Promise<void>;
}

export interface ProjectRepository {
  findById(id: string): Promise<Project | null>;
  findAll(filter?: ProjectFilter): Promise<Project[]>;
  findByClientId(clientId: string): Promise<Project[]>;
  create(project: Project): Promise<Project>;
  update(project: Project): Promise<Project>;
  delete(id: string): Promise<void>;
}

export interface SpecificationRepository {
  findById(id: string): Promise<Specification | null>;
  findByProjectId(projectId: string): Promise<Specification | null>;
  create(spec: Specification): Promise<Specification>;
  update(spec: Specification): Promise<Specification>;
  delete(id: string): Promise<void>;
}

export interface QuestionRepository {
  findById(id: string): Promise<Question | null>;
  findAll(): Promise<Question[]>;
  findBySection(section: string): Promise<Question[]>;
  create(question: Question): Promise<Question>;
  update(question: Question): Promise<Question>;
  delete(id: string): Promise<void>;
}

export interface QuestionnaireResponseRepository {
  findById(id: string): Promise<QuestionnaireResponse | null>;
  findByProjectId(projectId: string): Promise<QuestionnaireResponse[]>;
  findByRespondentId(respondentId: string): Promise<QuestionnaireResponse[]>;
  create(response: QuestionnaireResponse): Promise<QuestionnaireResponse>;
  update(response: QuestionnaireResponse): Promise<QuestionnaireResponse>;
  delete(id: string): Promise<void>;
}

export interface InvoiceRepository {
  findById(id: string): Promise<Invoice | null>;
  findByProjectId(projectId: string): Promise<Invoice[]>;
  findByClientId(clientId: string): Promise<Invoice[]>;
  create(invoice: Invoice): Promise<Invoice>;
  update(invoice: Invoice): Promise<Invoice>;
  delete(id: string): Promise<void>;
}

export interface TaskRepository {
  findById(id: string): Promise<Task | null>;
  findAll(filter?: TaskFilter): Promise<Task[]>;
  findBySprintId(sprintId: string): Promise<Task[]>;
  create(task: Task): Promise<Task>;
  update(task: Task): Promise<Task>;
  delete(id: string): Promise<void>;
}

export interface SprintRepository {
  findById(id: string): Promise<Sprint | null>;
  findByProjectId(projectId: string): Promise<Sprint[]>;
  findActive(projectId: string): Promise<Sprint | null>;
  create(sprint: Sprint): Promise<Sprint>;
  update(sprint: Sprint): Promise<Sprint>;
  delete(id: string): Promise<void>;
}

export interface NotificationRepository {
  findById(id: string): Promise<Notification | null>;
  findByUserId(userId: string): Promise<Notification[]>;
  findUnreadByUserId(userId: string): Promise<Notification[]>;
  countUnreadByUserId(userId: string): Promise<number>;
  create(notification: Notification): Promise<Notification>;
  markAsRead(id: string): Promise<void>;
  markAllAsRead(userId: string): Promise<void>;
  delete(id: string): Promise<void>;
}

export interface MessageRepository {
  findById(id: string): Promise<Message | null>;
  findByThreadId(threadId: string): Promise<Message[]>;
  create(message: Message): Promise<Message>;
  update(message: Message): Promise<Message>;
  delete(id: string): Promise<void>;
}

export interface ThreadRepository {
  findById(id: string): Promise<Thread | null>;
  findAll(): Promise<Thread[]>;
  findByProjectId(projectId: string): Promise<Thread[]>;
  findByParticipantId(participantId: string): Promise<Thread[]>;
  create(thread: Thread): Promise<Thread>;
  update(thread: Thread): Promise<Thread>;
  delete(id: string): Promise<void>;
}

export interface RoleRepository {
  findById(id: string): Promise<RBACRole | null>;
  findByName(name: string): Promise<RBACRole | null>;
  findAll(): Promise<RBACRole[]>;
  create(role: RBACRole): Promise<RBACRole>;
  update(role: RBACRole): Promise<RBACRole>;
  delete(id: string): Promise<void>;
}
